### PR TITLE
Run `.devcontainer/launch.sh` with bash + add error checking

### LINKED ISSUE #406: [BUG]: devcontainer/launch.sh fails if host shell is not bash
### Is this a duplicate?

- [X] I confirmed there appear to be no [duplicate issues](https://github.com/NVIDIA/cccl/issues) for this bug and that I agree to the [Code of Conduct](CODE_OF_CONDUCT.md)

### Type of Bug

Something else

### Component

Not sure

### Describe the bug

I use zsh, not bash. Running `.devcontainer/launch.sh` produces syntax warnings and then continues to launch an (incorrectly configured) vscode.

### How to Reproduce

```
$ /bin/zsh
$ .devcontainer/launch.sh
.devcontainer/launch.sh: 7: Bad substitution
.devcontainer/launch.sh: 9: [[: not found
.devcontainer/launch.sh: 9: [[: not found
cp: cannot stat '.../src/rapids/first-party/.devcontainer/cuda-/devcontainer.json': No such file or directory
sed: can't read /tmp/tmp.kp1eIGeWeO/first-party/.devcontainer/devcontainer.json: No such file or directory
devcontainer URL: vscode://vscode-remote/dev-container+2f746d702f746d702e6b70316549476557654f2f66697273742d7061727479/home/coder/cccl
```

### Expected behavior

This should work, and run in `/bin/bash`.

### Reproduction link

_No response_

### Operating System

_No response_

### nvidia-smi output

_No response_

### NVCC version

_No response_

### PR DESCRIPTION
**Pull request by wence-**
## Description

For the shebang line to work, it must be the first line of the file (not the second as at present).

While here, add `set -euo pipefail` and notice, and provide, a useful error message if we've requested a CTK + compiler combo that doesn't have a prebaked devcontainer.json. Previously, we would get errors from `cp` and `sed` and then launch vscode anyway.

- Closes #406

A partial step towards #126, in that unknown devcontainer combinations are now at least noticed.

## Checklist

- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

### CODE REVIEW

File: .devcontainer/launch.sh [hunk 1]
```diff
@@ -1,23 +1,28 @@
...
     if [[ -z $1 ]] || [[ -z $2 ]]; then
         echo "Usage: $0 [CUDA version] [Host compiler]"
-        echo "Example: $0 12.1 gcc12"
+        echo "Example: $0 12.2 gcc12"
         return 1
     fi
+    set -euo pipefail
+    # Ensure we're in the repo root
+    cd "$( cd "$( dirname "$(realpath -m "${BASH_SOURCE[0]}")" )" && pwd )/..";
+
```
Comment (miscco) [inline .devcontainer/launch.sh:14]: nitpick:
```suggestion

    set -euo pipefail
    # Ensure we're in the repo root
    cd "$( cd "$( dirname "$(realpath -m "${BASH_SOURCE[0]}")" )" && pwd )/..";
```

**Review by miscco — APPROVED** [review]

**Comment by miscco** [comment]
Thanks a lot for the contribution 🎉

**Review by jrhemstad — APPROVED** [review]
Thanks @wence-! 

Admittedly, I have been neglecting the `launch.sh` ever since I hit a roadblock in supporting custom CTK/host compiler combinations. I wrote up the status quo here: https://github.com/NVIDIA/cccl/issues/126#issuecomment-1708445795

**Comment by miscco** [comment]
@wence- could you please add the `[skip-tests]` tag to your commit message.

We want to avoid running expensive GPU tests for pure infrastructure fixes. Any issues with the changes will be found by the CPU-only build tests

**Comment by wence-** [comment]
> could you please add the `[skip-tests]` tag to your commit message.

Done, thanks!

**Comment by miscco** [comment]
Thanks a lot for fixing the launch script 🥇

**Pull request merged**